src/zzip.cpp

```diff
@@ -0,0 +1,882 @@
+                                  content_end
+                              );
+            if( compressed_size == 0 ) {
+                return nullptr;
+            }
+        }
+        new_entries.emplace_back( compressed_entry{ std::move( file_relative_path ), content_end, compressed_size } );
+        bytes_written += compressed_size;
+        input_consumed += file->len();
+        content_end += compressed_size;
+        raw_input_remaining -= file->len();
+    }
+    zip->update_footer( original_footer, content_end, new_entries, /* shrink_to_fit = */ true );
+    return zip;
+}
+
+bool zzip::extract_to_folder( const std::filesystem::path &path,
+                              const std::filesystem::path &folder,
+                              const std::filesystem::path &dictionary )
+{
+    std::shared_ptr<zzip> zip = zzip::load( path, dictionary );
+
+    for( const JsonMember &entry : zip->footer_.get_object( kEntriesKey ) ) {
+        std::filesystem::path filename = std::filesystem::u8path( entry.name() );
+        size_t len = zip->get_file_size( filename );
+        std::filesystem::path destination_file_path = folder / filename;
+        if( !assure_dir_exist( destination_file_path.parent_path() ) ) {
+            return false;
+        }
+        std::shared_ptr<mmap_file> file = mmap_file::map_writeable_file( destination_file_path );
+        if( !file || !file->resize_file( len ) ) {
+            return false;
+        }
+        if( zip->get_file_to( filename, static_cast<std::byte *>( file->base() ), file->len() ) != len ) {
+            return false;
+        }
+    }
+    return true;
+}
+
+bool zzip::compact( double bloat_factor )
+{
+    zzip_footer footer{ footer_ };
+    std::optional<zzip_meta> meta_opt = footer.get_meta();
+    if( !meta_opt.has_value() ) {
+        return false;
+    }
+
+    zzip_meta &meta = *meta_opt;
+    size_t max_allowed_size = meta.total_content_size * bloat_factor + 1024;
+    // Are we even going to try to compact?
+    if( file_->len() < max_allowed_size ) {
+        return false;
+    }
+
+    // Can we save at least one physical page?
+    size_t delta = file_->len() - ( meta.total_content_size + 1024 );
+    if( delta < 4 * 1024 ) {
+        return false;
+    }
+
+    std::vector<compressed_entry> compressed_entries = footer.get_entries();
+
+    // If we were compacting in place, sorting left to right would ensure that we
+    // only move entries left that we need to move. Unchanged data would stay on the left.
+    // However that exposes us to data loss on power loss because zzip recovery
+    // relies on an intact sequence of zstd frames.
+    std::sort( compressed_entries.begin(), compressed_entries.end(), []( const compressed_entry & lhs,
+    const compressed_entry & rhs ) {
+        return lhs.offset < rhs.offset;
+    } );
+
+    std::filesystem::path tmp_path = path_;
+    tmp_path.replace_extension( ".zzip.tmp" ); // NOLINT(cata-u8-path)
```

Comment:
Shouldn't you use `mkstemp` here?

Reply:
The old save path always wrote the .tmp file adjacent to its replacement.

Reply:
Yea, that’s why I suggested `mkstemp` rather than `std::tmpfile`; it lets you specify the path. It's not so much about security (in this case) as it is about avoiding a filename that any other process is using for any reason, such as Dropbox. As an alternative, consider calling `open` with the `O_TMPFILE` flag on Linux. This gives you an unnamed file that Dropbox can’t see until you name it in the final step; I assume there’s something similar on other platforms.

Reply:
Ok, on Windows std::filesystem::rename is implemented with MoveFileEx with MOVEFILE_REPLACE_EXISTING which is atomic (enough) on a same-volume move https://github.com/microsoft/STL/blob/main/stl/src/filesystem.cpp#L722. The linux rename api is guaranteed atomic.

The issue with Dropbox isn't that it's using the filename, it's that it briefly locks the file after it is closed before the rename so it can index it.

Reply:
Right, the point is to prevent it from being able to do that.

Reply:
There's not really a way to do that which preserves atomicity on all platforms. There's no guarantee we have a writeable location on the same volume that isn't watched by another process.

Reply:
I tested this code a bunch and for whatever reason, the zzip codepaths do not trigger issues with dropbox, despite repeatedly having issues with eg. map memory or other misc files in the root of the save folder. I never hit a case of needing to retry the save here.